### ISSUE TITLE
Multi-context support for C extensions

### ISSUE THREAD

**Issue by timfel** [oracle/graalpython#437]
### TL;DR

Native extension libraries such as for popular DS and ML packages cannot be loaded multiple times into the same process, precluding their use in embeddings that use multiple GraalPy contexts. We will try to alleviate this issue for all native extensions.

### Goals

The primary goal for allowing native extensions in multiple contexts is to enable the scaling of GraalPy embeddings. While each GraalPy context is subject to the Python global interpreter lock (GIL), multiple GraalPy contexts can be used to execute Python code in parallel on different Java threads while sharing compiled code caches. If we managed to allow native extensions to be loaded multiple times and used in different contexts, this scaling strategy could be extended to the vast majority of Python packages.

### Non-Goals

- **Complete isolation between native extensions**. The goal is to be able to load and run native extensions across contexts, but native code will have full access to the process memory and resources. Thus, ill-behaved native extensions can still affect other GraalPy contexts.

**Comment by msimacek** [comment]
Hi @shenzhaoquan, there is already experimental support this in 24.2 release. See https://github.com/oracle/graalpython/blob/master/docs/user/Native-Extensions.md#multi-context-and-native-libraries

**Comment by HolgerHees** [comment]
Is there any other solutions planned? Copy the libraries works only in tiny scenarios. But not in a huge application where you work with a lot of different contexts and each of them could theoretically use native modules.

e.g. in our application we provide the possibility for users to automate everything with python scripts. Each script works with his own context and could theoretically use native modules like "requests". Normally I would install "requests" via pip and make this available for all script use cases.

With your solution, I have to create for each script an individual venv and copy all libs to be prepared for the usage of e.g. "requests"

btw. with jython I was never faced with this kind of problem. In each jython context I was able to use modules like "requests"

**Comment by msimacek** [comment]
The copying is automated and should work transparently, there's no need for you to copy anything manually. The limitation is that you need to have an upper bound on the number of contexts you create.

`requests` is not a native module, it just uses some native module in the standard library, which is something we can probably fix by reimplementing that module in Java. Then you wouldn't need the native module isolation. Jython never supported running native modules at all.

Jython has no context isolation. Jython "contexts" are just different namespaces for global variables, but the modules and their state are shared. The equivalent on GraalPy is to just use a single context and exec the code with new globals. Like this:
```
Value runFn = context.eval("python", "lambda code: exec(code, {})");
runFn.execute("a = 1");
```

**Comment by tulinski-rtbh** [comment]
Hello, just checking in to see if this is currently on the roadmap or if there's an estimated timeline for a fix? Thanks!
